### DISCUSSION TITLE
Server build import with particular setup

### DISCUSSION THREAD

**Discussion by alexturpin** [vikejs/vike#797] [Discussion] [answered]
Hey everyone,

Finally adding VPS to my project, and running into issues with the server build import. When I hit a route that the vps middleware should handle, I get:

```
Error: [@brillout/vite-plugin-import-build@0.2.16][Wrong Usage] Cannot find server build. (Re-)build your app and try again. If you still get this error, then you may need to manually import the server build, see https://github.com/brillout/vite-plugin-import-build#manual-import
```

I've read that page and both linked pages but am still a bit unsure on how to best resolve it for my particular setup.

The way my app is setup, I've got a CloudFlare worker in `worker.ts` (which is just a HatTip server), which is built using esbuild through a Vite plugin. Then, I use Miniflare in the same Vite plugin to run the server and get all the Cloudflare APIs locally.

My Vite plugin that does the esbuild for the server outputs the worker file in `config.build.outDir + "/worker.mjs"`, which was typically `dist/worker.mjs`. However I notice that when including vps as a Vite plugin, the output becomes `dist/client/worker.mjs`.  I tried making it explicitly both `dist/server/worker.mjs` and `dist/server/index.js` but neither seem to get picked up properly by vps.

I've seen that one solution is to manually import it but I'm quite confused as to where I should do that.`worker.ts` or `server.ts` (where my HatTip server lives) doesn't seem to make a whole lot of sense as that would basically be a recursive import in a way.

I realize that this setup is pretty specific, but I also know that's one of the strengths of VPS and why I picked it. I'm hoping someone has some suggestions for how I can debug this further. In the meantime I'll see what I can do to create a minimum example that reproduces my problem.

Thank you 🙏

**Comment by brillout** [comment]
AFAICT it should just work. So that's surprising.

Set the `DEBUG` flag to `true` at line `3` of `node_modules/@brillout/vite-plugin-import-build/dist/shared/debugLogs.js` and let me know what you get.


> I tried making it explicitly both `dist/server/worker.mjs` and `dist/server/index.js`

Yes: every entry needs it. 

> but neither seem to get picked up properly by vps.

That's unexpected as well. What error do you get? What's the content of the file `dist/server/importBuild.cjs` you're importing?

> that's one of the strengths of VPS and why I picked it

Most definitely, VPS should be flexible in that regard as well.

**Comment by alexturpin** [comment]
Thanks for the lightning fast reply, one more reason why this project rocks 🔥 

Oh interesting, no `importBuild.cjs` file is getting generated at all, only worker.mjs (the one I'm building with my esbuild plugin). I hadn't realized that's what needed to be imported, that's likely the culprit... I must be circumventing some Vite thing that generates that with my weird build setup 😅 

Definitely sounds like a "me" problem and not VPS, and it sounds like I've got some reading to do on `importBuild.cjs`. Here's my minimum repro either way, in case you are curious: https://github.com/alexturpin/vps-miniflare/tree/main

**Reply by brillout** [reply]
If you bundle your server build then it's expected (and supported) that you don't need the `importBuild.cjs` file after bundling.

The question is where and what `importBuild.cjs` is *before* bundling.

**Reply by brillout** [reply]
> Thanks for the lightning fast reply, one more reason why this project rocks 🔥

Btw. in case you think you can (eventually) convince your employer: https://github.com/sponsors/brillout 🙂.

**Reply by alexturpin** [reply]
Commenting out my little cloudflare vite plugin does make VPS work as expected. I'm somewhat unfamiliar with `importBuild.cjs` and am not finding much online other than VPS doc pages. Would you have any more resources where I could better understand what the file is so I can figure out how to insert that into my weird setup?

**Reply by alexturpin** [reply]
@brillout I'm a founder building a startup and you are at the top of the list when we generate revenue, and I'll definitely send a single (personal) contribution your way if I can figure out this setup now!

**Reply by brillout** [reply]
https://github.com/brillout/vite-plugin-import-build
https://github.com/brillout/vite-plugin-import-build/issues/4

**Reply by alexturpin** [reply]
So importBuild.cjs isn't a Vite generated thing?

**Reply by brillout** [reply]
> @brillout I'm a founder building a startup and you are at the top of the list when we generate revenue, and I'll definitely send a single contribution your way if I can figure out this setup now!

Thanks, I appreciate it!

**Reply by brillout** [reply]
> So importBuild.cjs isn't a Vite generated thing?

No it's generated by https://github.com/brillout/vite-plugin-import-build.

**Reply by alexturpin** [reply]
Ahh. Running `vite build` before `vite dev` seems to make it work, so I'm assuming my plugin is missing a chunk that should be telling Vite (which I assume then tells VPS?) to build that part

**Reply by brillout** [reply]
> Ahh. Running `vite build` before `vite dev` seems to make it work, so I'm assuming my plugin is missing a chunk that should be telling Vite (which I assume then tells VPS?) to build that part

It's only needed in production: see https://github.com/brillout/vite-plugin-import-build/commit/36ceaf54b50b838e1285ddea0d9a80b05292588a. I've also improved the "How it works" ticket.

Let me know if you have more questions.

**Reply by alexturpin** [reply]
Thanks, the context is helpful. I'm not sure how to fix this in my case, as I was actually trying to run this in dev not in prod, but I think I just need to rebuild my weird setup from the ground up starting with VPS instead of adding it after, to figure out what I'm doing wrong.

**Reply by brillout** [reply]
I released `0.4.115` which includes new warnings that may help you.

Keep me updated.

**Reply by alexturpin** [reply]
Thanks for the new version. Haven't had any luck this afternoon. It seems like the vite-plugin-cloudflare example with VPS doesn't work either https://github.com/Aslemammad/vite-plugin-cloudflare/issues/25. I'll see if Mohammad has any idea. In the meantime I think my next step is to start from the basic VPS scaffold and try to add vite-plugin-cloudflare / Miniflare to that and see where things break. That order might make more sense.

**Reply by brillout** [reply]
Check the `/examples/cloudflare-*` examples (which are part of VPS's CI).

**Reply by alexturpin** [reply]
Those are very nice indeed but I'm trying to use Miniflare in dev (rather than Express) for the best possible DX, in order to be able to easily develop with KV, Durable Objects, etc. While in theory this should be possible with the HatTip Cloudflare adapter, this is why my problems are happening of course 🙃

**Reply by alexturpin** [reply]
If I do manage to get a simple example of VPS + HatTip + Miniflare it'll probably be worth contributing to those examples as IMO that would be just best in class DX for Cloudflare app development.

**Reply by brillout** [reply]
Sounds good 👍. As for HatTip: can you create a ticket? I believe Fatih had some ideas about Miniflare.

**Reply by alexturpin** [reply]
Certainly https://github.com/hattipjs/hattip/issues/64

Although to be clear, HatTip + Miniflare do work great together, I've had that in my repo for several weeks now. Just unsure how to make it all work with VPS. I suspect Miniflare itself isn't the culprit, but rather how I'm building the server code. I see the HatTip VPS example uses Vavite, perhaps that's the missing piece here.

**Reply by brillout** [reply]
> HatTip + Miniflare do work great together

Is there a reproduction that doesn't use vite-plugin-cloudflare? (I've already checked https://github.com/alexturpin/vps-miniflare/commits/main.)

I'd be up to dig into a minimal reproduction that uses HatTip + Miniflare + VPS.

**Reply by alexturpin** [reply]
I would absolutely love to be able to give you such a repro but that's kinda where I'm hitting a wall unfortunately haha. 

vite-plugin-cloudflare (or my modified version in alexturpin/vps-miniflare) bundles the server with esbuild and runs Miniflare as a Vite plugin middleware. This is probably as minimal as it gets if using esbuild to bundle the server.

The other alternative I can see would be using Vavite ([as per the HatTip + VPS example](https://github.com/hattipjs/hattip/tree/main/examples/vavite/vite-plugin-ssr)). It seems that Vavite's build process works great with VPS, but I'm unsure how to integrate Miniflare into it. I'm just looking into this avenue today, it looks like using `serverEntry` will be part of the solution to then be able to get a `new Miniflare({...})` server going, but TBD on whether I can get that working.

**Reply by brillout** [reply]
Why do you want to use Miniflare in the first place? Why not this https://vite-plugin-ssr.com/cloudflare-workers#development instead?

FYI HatTip recently released `@hattip/vite` (which is basically a small wrapper on top of Vavite).

**Reply by alexturpin** [reply]
Oh that's cool didn't know about HatTip/vite!

I want to use Miniflare because it emulates all of the Cloudflare APIs locally, such as KV (kinda like redis), D1 (database), etc. I think that's a big advantage of using Cloudflare, at this point they have APIs for almost everything you could need, from a key value store, queue, database, Cron, etc. all available locally with Miniflare for a seamless DX.

**Reply by brillout** [reply]
I see. So you actually can't easily reproduce the production environment without Miniflare, correct?

**Reply by brillout** [reply]
> I see. So you actually can't easily reproduce the production environment without Miniflare, correct?

For future readers: that is correct. (We've been PM'ing on Discord.)

We're working on a Miniflare example.

**Reply by alexturpin** [reply]
I haven't had enough time to play with this this afternoon but will absolutely report back next week when I get it working :) thanks to Rom for all the help!

**Reply by brillout** [reply]
Looking forward to it. And thanks for the donation!

**Comment by alexturpin** [comment]
Just to give a small update. The current path I'm attempting is:

1. Remove the cloudflare plugin from vite.config.ts
2. Add the import `../../dist/server/importBuild.cjs` to `/src/server/worker.ts`
3. Apply esbuild directly on `/src/server/worker.ts`

For now I'm foregoing any HatTip business and just trying to get VPS to run through Miniflare.

So far that hasn't been very successful. With the most basic setup https://github.com/alexturpin/vps-mf, Miniflare balks with `Error: Dynamic require of "tty" is not supported` which seems to be coming from VPS' `picocolors` dependency.

**Reply by alexturpin** [reply]
One approach I could take at this point would be to completely separate my API (running through Miniflare) from my vite-plugin-ssr setup, probably using a monorepo and deploying them as two separate workers. I was hoping to be able to leverage Miniflare directly in VPS' SSR but that seems more difficult to setup than is worth, and I suppose I could always make a fetch request to my API from VPS' SSR as needed.

**Reply by brillout** [reply]
> Miniflare balks with `Error: Dynamic require of "tty" is not supported`

Have you tried to bundle the whole server? Does Miniflare expect the whole server to be bundled into a single file like Cloudflare Workers?

**Reply by brillout** [reply]
Ah, actually, `tty` is a Node.js module. If you remove the `tty` import from the bundle, does it work then? Picocolors is a fairly stable dependency; a fork would be easy to maintain.

**Reply by alexturpin** [reply]
Yeah killing a few more things from `picocolor` does get it past those errors.

**Reply by alexturpin** [reply]
Now I'd be at the point of importing `importBuild.cjs` but one thing I'm not sure of is... it doesn't seem to be generated in dev right? How should I import it in dev?

**Reply by alexturpin** [reply]
It is starting to feel more and more like using separate workers would be the best way to make this work canonically with VPS though. Especially seeing as I seem to be the first / only one trying to run VPS through Miniflare.

**Reply by brillout** [reply, marked as answer]
> Now I'd be at the point of importing `importBuild.cjs` but one thing I'm not sure of is... it doesn't seem to be generated in dev right? How should I import it in dev?

You need `importBuild.cjs` only for loading the build files at `dist/server/`. In dev, you don't need `importBuild.cjs` and instead it's Vite's development server that crawls and dynamically imports the user land files.

So it's either you run `$ vite build` and import `importBuild.cjs`, or you instantiate Vite's development server as described [here](https://vite-plugin-ssr.com/renderPage).

> It is starting to feel more and more like using separate workers would be the best way to make this work canonically with VPS though. 

I believe the question here is whether Vite's development server is able to run in Miniflare? If Miniflare cannot even import `tty`, then the chances are slim for running Vite in Miniflare.

If Vite can't run in Miniflare and you want Vite's dev DX such as HMR then, yes, you'll have to run Vite separately from Miniflare.

**Reply by alexturpin** [reply]
Right so I was running Miniflare directly, not through Vite's dev server, as I was understand was one of the suggestions for how to get this working. But indeed it seems like it will be way easier to just run these two things independently, and deploy them as two separate workers.
